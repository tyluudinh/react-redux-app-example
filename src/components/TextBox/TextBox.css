.custom-textbox {
  width: 100%;
  height: 50px;
  padding: 10px 17px;
  line-height: 30px;
  font-size: 0.8rem;
  font-weight: lighter;

  border: 0;
  box-shadow: 0 1px 3px rgba(0,0,0,0.09), 0 1px 2px rgba(0,0,0,0.16);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  &:focus {
    outline: none;
    box-shadow: 0 2px 4px rgba(0,0,0,0.25), 0 4px 6px rgba(0,0,0,0.22);
  }
  &.multiline {
    height: 100px;
  }
}
input.custom-textbox::-webkit-input-placeholder {
  color: #3d3d3d;
  opacity: 0.3;
}

.custom-textbox-wrapper {
  position: relative;
  &.with-icon {
    padding-left: 50px;
  }

  .icon-wrapper {
    position: absolute;
    left: 0;
    top: 0;
    bottom: 0;
    width: 50px;
    height: 50px;
    text-align: center; 
    line-height: 50px;
    background: rgba(0, 43, 83, 0.3);
  }
  .icon-wrapper__icon {
    fill: white;
  }

  /** reverse color mode, for colored background */
  &.reverse {
    .custom-textbox{
      box-shadow: none;
      background: rgba(0, 0, 0, 0.25);
      color: white;
    }
    .custom-textbox:hover{
      box-shadow: none;
      color: white;
    }
    input {
      font-size: 0.8rem;
    }
    input.custom-textbox::placeholder {
      color: white;
    }
  }
}

.without-boxshadow {
  .custom-textbox {
    box-shadow: none;
  }
  .custom-textbox:hover {
    box-shadow: none;
  }
}

/* Phone number input styles */

.custom-textbox-wrapper {
  .rrui__input {
    height: 50px !important;
    background:  white;
  }
  .rrui__select__button {
    border-bottom: 0;
    background: #dedede;
  }
  .rrui__select__selected-label {
    padding: 0 12px;
    display: flex;
  }
  .rrui__select__arrow {
    display: none;
  }
  .react-phone-number-input__country.rrui__select--collapsed {
    margin-right: 0;
  }
  input.react-phone-number-input__phone {
    border-bottom: 0;
    font-weight: lighter;
    padding-left: 12px;
    box-shadow: 0 1px 3px rgba(0,0,0,0.09), 0 1px 2px rgba(0,0,0,0.16);
    transition: all 0.3s cubic-bezier(.25,.8,.25,1);
    font-size: 0.8rem;
    &:focus {
      outline: none;
      box-shadow: 0 2px 4px rgba(0,0,0,0.25), 0 4px 6px rgba(0,0,0,0.22);
    }
  }

  &.reverse {
    .rrui__select__button {
      background: transparent;
      svg {
        fill: white;
      }
    }
    .rrui__input {
      background: rgba(0, 0, 0, 0.25);
    }
    input.react-phone-number-input__phone {
      color: white;
      font-size: 0.8rem;
      box-shadow: none;
      padding-left: 0;
      &:focus {
        box-shadow: none;
      }
    }
    input.rrui__input.rrui__input-element.rrui__input-field.react-phone-number-input__phone::placeholder {
      color: white !important;
      padding-left: 0;
    }

  }
}