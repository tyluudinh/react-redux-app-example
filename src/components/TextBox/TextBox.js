import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

// React Phone Number input
import Phone from 'react-phone-number-input';
import 'react-phone-number-input/rrui.css';
import 'react-phone-number-input/style.css';

import './TextBox.css';

export default class TextBox extends Component {
  constructor(props) {
    super(props);

    this.textBoxOnChange = this.textBoxOnChange.bind(this);
    this._renderInput = this._renderInput.bind(this);
  }

  textBoxOnChange(e) {
    const { onChange, name } = this.props;
    // pass the value to outer event handler
    // If target property doesn't exist => special input field (value return directly)
    if (e) {
      onChange(name, (e.target) ? e.target.value : e);
    }
  }

  _renderInput() {
    const { type, name, value, placeholder, multiline } = this.props;
    if (type === 'phone') {
      return (
        <Phone 
          placeholder="Phone number"
          value={ value }
          onChange={this.textBoxOnChange}
          />
      )
    }
    if (multiline) {
      return (
        <textarea
          className={classNames('custom-textbox multiline')}
          name={name}
          value={value}
          onChange={this.textBoxOnChange}
          placeholder={placeholder}
        ></textarea>
      )
    }
    return (
      <input 
        type={(type) ? type : 'text'}
        className={classNames('custom-textbox',)}
        name={name}
        value={value}
        onChange={this.textBoxOnChange}
        placeholder={placeholder}
      />
    )
  }

  _renderIcon(withIcon) {
    const { name, style, size } = withIcon;
    const Icon = require(`react-icons/lib/fa/${name}`);
    if (Icon) {
      return (
        <div className="icon-wrapper">
          <Icon style={(style) ? style : {}} size={(size) ? size : null} className="icon-wrapper__icon" />
        </div>
      )
    }
    // Ignore 
    // It's hard to reproduce this icon missing case
    // istanbul ignore next
    return null;
  }

  render() {
    const { className, withIcon, reverseTheme } = this.props;

    return (
      <div 
        className={
          classNames(
            'custom-textbox-wrapper',
            { 
              'with-icon': withIcon,
              'reverse': reverseTheme,
            },
            className,
          )
        }
      >
        {(withIcon) ? this._renderIcon(withIcon) : null}
        {this._renderInput()}
      </div>
    )
  }
}

TextBox.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  className: PropTypes.string,
  placeholder: PropTypes.string,
  withIcon: PropTypes.object,
  name: PropTypes.string,
  reverserTheme: PropTypes.bool,
  type: PropTypes.string,
  multiline: PropTypes.bool,
};