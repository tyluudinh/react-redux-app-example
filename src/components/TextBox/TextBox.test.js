import React from 'react';

import TextBox from './TextBox';

describe('Component TextBox', () => {
  const mockOnChange = jest.fn();
  const getTextBox = (props) => mount(<TextBox {...props} />)
  test('Render correctly', () => {
    // Normal Textbox 
    const normalTextBoxProps = {
      onChange: mockOnChange,
      placeholder: 'My name is',
      value: 'Poe'
    };
    const normalTextBox = getTextBox(normalTextBoxProps);
    const normalTextBoxJson = toJson(normalTextBox);
    expect(normalTextBox.find('input').prop('type')).toBe('text');
    expect(normalTextBox.find('input').prop('placeholder')).toBe('My name is');
    expect(normalTextBoxJson).toMatchSnapshot();

    // Custom type, with icon and reverse theme TextBox
    const customType = 'password';
    const customizedTextBoxProps = {
      withIcon: 'lock',
      className: 'ucustom-class',
      reverseTheme: true,
      type: customType,
      value: 'Poe',
      onChange: mockOnChange,
    };
    const customizedTextBox = getTextBox(customizedTextBoxProps);
    // Get the root component
    const customizedTextBoxRoot = customizedTextBox.find('div').at(0);
    const customizedTextBoxJson = toJson(customizedTextBox);
    expect(customizedTextBox.find('input').prop('type')).toBe(customType)
    expect(customizedTextBoxRoot.hasClass('reverse')).toBeTruthy();
    expect(customizedTextBoxRoot.hasClass('with-icon')).toBeTruthy();
    expect(customizedTextBoxJson).toMatchSnapshot();
  })

  describe('Event handler - onChange', () => {
    test('Called correctly', () => {
      const component = getTextBox({
        value: 'Poe',
        onChange: mockOnChange,
      });
      component.find('input').simulate('change');
      expect(mockOnChange.mock.calls.length).toBe(1);
    })

    test('Called with correct param', () => {
      const props = {
        value: 'Poe',
        onChange: mockOnChange,
        name: 'myname'
      };
      const newValue = 'new Poe';
      const fakeEventTarget = { name: props.name, value: newValue };
      const component = getTextBox(props);
      component.find('input').simulate('change', { target: fakeEventTarget });
      expect(mockOnChange.mock.calls[1][0]).toBe(props.name);
      expect(mockOnChange.mock.calls[1][1]).toBe(newValue);
    })
  })
});