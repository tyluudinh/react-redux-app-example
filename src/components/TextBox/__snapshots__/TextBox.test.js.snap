// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component TextBox Render correctly 1`] = `
<TextBox
  onChange={[Function]}
  placeholder="My name is"
  value="Poe"
>
  <div
    className="custom-textbox-wrapper"
  >
    <input
      className="custom-textbox"
      name={undefined}
      onChange={[Function]}
      placeholder="My name is"
      type="text"
      value="Poe"
    />
  </div>
</TextBox>
`;

exports[`Component TextBox Render correctly 2`] = `
<TextBox
  className="ucustom-class"
  onChange={[Function]}
  reverseTheme={true}
  type="password"
  value="Poe"
  withIcon="lock"
>
  <div
    className="custom-textbox-wrapper with-icon reverse ucustom-class"
  >
    <div
      className="icon-wrapper"
    >
      <FaLock
        className="icon-wrapper__icon"
      >
        <IconBase
          className="icon-wrapper__icon"
          viewBox="0 0 40 40"
        >
          <svg
            className="icon-wrapper__icon"
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style={
              Object {
                "color": undefined,
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 40 40"
            width="1em"
          >
            <g>
              <path
                d="m14.1 17.1h11.5v-4.2q0-2.4-1.7-4.1t-4-1.7-4.1 1.7-1.7 4.1v4.2z m18.6 2.2v12.8q0 0.9-0.6 1.6t-1.5 0.6h-21.5q-0.8 0-1.5-0.6t-0.6-1.6v-12.8q0-0.9 0.6-1.5t1.5-0.7h0.8v-4.2q0-4.1 2.9-7.1t7.1-2.9 7 2.9 3 7.1v4.2h0.7q0.9 0 1.5 0.7t0.6 1.5z"
              />
            </g>
          </svg>
        </IconBase>
      </FaLock>
    </div>
    <input
      className="custom-textbox"
      name={undefined}
      onChange={[Function]}
      placeholder={undefined}
      type="password"
      value="Poe"
    />
  </div>
</TextBox>
`;
